pkg/utils/overriding/keys.go

```diff
@@ -6,138 +6,33 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 )
 
-type keyedCommand struct {
-	Id string `json:"id,omitempty"`
-
-	workspaces.Command `json:",inline"`
-}
-
-type keyedComponent struct {
-	Name string `json:"name,omitempty"`
-
-	workspaces.Component `json:",inline"`
-}
-
-type keyedDevWorkspaceTemplateSpecContent struct {
-	// Projects worked on in the workspace, containing names and sources locations
-	// +optional
-	Projects []workspaces.Project `json:"projects,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
-
-	// List of the workspace components, such as editor and plugins,
-	// user-provided containers, or other types of components.
-	// +optional
-	Components []keyedComponent `json:"components,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
-
-	// Predefined, ready-to-use, workspace-related commands.
-	// +optional
-	Commands []keyedCommand `json:"commands,omitempty" patchStrategy:"merge" patchMergeKey:"id"`
-
-	// Bindings of commands to events.
-	// Each command is referred-to by its name.
-	// +optional
-	Events *workspaces.Events `json:"events,omitempty"`
-}
-
-func addKeys(keyedSpec *keyedDevWorkspaceTemplateSpecContent) error {
-	for idx := range keyedSpec.Commands {
-		key, err := keyedSpec.Commands[idx].Key()
-		if err != nil {
-			return err
-		}
-		keyedSpec.Commands[idx].Id = key
-	}
-	for idx := range keyedSpec.Components {
-		key, err := keyedSpec.Components[idx].Key()
-		if err != nil {
-			return err
-		}
-		keyedSpec.Components[idx].Name = key
-	}
-	return nil
-}
-
-func removeKeys(keyedSpec *keyedDevWorkspaceTemplateSpecContent) *workspaces.DevWorkspaceTemplateSpecContent {
-	result := workspaces.DevWorkspaceTemplateSpecContent{
-		Projects: keyedSpec.Projects,
-		Events:   keyedSpec.Events,
-	}
-
-	if keyedSpec.Commands != nil && len(keyedSpec.Commands) > 0 {
-		result.Commands = []workspaces.Command{}
-	}
-
-	if keyedSpec.Components != nil && len(keyedSpec.Components) > 0 {
-		result.Components = []workspaces.Component{}
-	}
+type checkFn func(elementType string, keysSets []sets.String) []error
```

Comment:
There is an implicit assumption here that the `keysSets` passed to `checkFn` will maintain some sort of order from the `...workspaces.TopLevelListContainer` passed in to `checkKeys`

Reply:
Is it clearer with the documentation provided in commithttps://github.com/devfile/api/pull/122/commits/f5bcbc623c4b0631f46e1bacea35ac477cadf610 ?

Reply:
It makes more sense, but I still think the structure of `checkFn` and `checkKeys` is strange -- `checkKeys` is doing too much. It's effectively

1. Merging a `[]map[string][]string` into `map[string][][]string` based on map keys
2. Applying `checkFn` to each `(string, []string)` pair in the flattened map to map the `[]string` to error

While this is a very general function, it's also not a "standard" function one would expect to see, so understanding how it works requires additional reading. While more verbose, I think splitting it into two functions would make it easier ot understand:

```go
func mergeTopLevelLists(...map[string][]string) map[string][]sets.String {...}

func checkTopLevelLists(checkFn, map[string][]sets.String) error {...}
```

Restricting it further based on current uses (so maybe not valid if I'm missing a case), I would go so far as to remove `checkTopLevelLists` and `checkFn` altogether, since the `elementType` argument for `checkFn` is only used for error messages:

```go
func ensureNoConflictWithParent(...) []error {
  mergedTopLevelLists := mergetopLevelLists(mainContent, parentflattenedContent)
  for elementType, listElem := range mergedTopLevelLists {
    mainKeys := listElem[0]
    parentOrPluginKeys := listElem[1]
    overriddenElementsInMainContent := mainKeys.Intersection(parentOrPluginKeys)
    if overriddenElementsInMainContent.Len() > 0 {
      return []error{fmt.Errorf("Some %s are already defined in parent: %s. "+
        "If you want to override them, you should do it in the parent scope.",
        elementType,
        strings.Join(overriddenElementsInMainContent.List(), ", "))}
    }
    return []error{}
  }
}
```
it's a bit more repetition (since we have to write the loop over `(string, [][]string`) pairs, but i would argue easier to understand. We could go further in this case, and iterate over key-value pairs in `mainContent` directly, to make our check more explicit.

This also raises a question about the current implementation: if we're to assume that `keysSets[0]` is potentially special (e.g. it is the `mainContent` keys above), how do we detect the difference between the case where a subsequent element gets merged down and introduces a `listType` not present in `mainContent`? For example, with `ensureNoConflictsWithPlugins`, if `mainContent` doesn't contain any `Components`, won't it seem like the first `pluginFlattenedContent`'s components represent components in `mainContent`? Perhaps more clearly, the code can't distinguish between
```
topLevelListContainer1 {
  "Commands": "hello"
  "Components": "I'm a component"
}
topLevelListContainer2 {
  "Commands": "hello2"
}
```
and 
```
topLevelListContainer1 {
  "Commands": "hello"
}
topLevelListContainer2 {
  "Commands": "hello2"
  "Components": "I'm a component"
}
```
since they both merge to
```
map[string][]sets.String {
  "Commands": [{"hello"}, {"hello2"}],
  "Components": [{"I'm a component"}],
}
```

Reply:
I suppose the last concern is a non-issue, because of they way `keyed_definitions.go` is written, but we're still relying on an implicit implementation in another package here.